students/will_chang/lesson04/mailroom_update.py

```diff
@@ -0,0 +1,73 @@
+    print("\n")
+    print(format_string_header.format(*("Donor Name", "Total Given", "Num Gifts", "Average Gift")))
+    print("-"*(max_len_str[0]+max_len_str[1]+max_len_str[2]+max_len_str[3]+38))
+    for row in initial_donor_list:
+        print(format_string.format(*row.values()))
+
+def thank_you():
+    """Prompts user to input new donor information."""
+    ask_name = input("\nPlease enter a full name or type 'list' for a list of the donor names: ")
+    name_in_list = False
+    while ask_name.lower() == "list":
+        print([donor["name"] for donor in initial_donor_list])
+        ask_name = input("\nPlease enter a full name or type 'list' for a list of the donor names: ")
+    for donor in initial_donor_list:
+        if ask_name == donor["name"] : # Check if the entered name is already in the list
+            name_in_list = True
+            ask_donor = float(input("Please enter a donation amount: "))
+            donor["total_don"] += ask_donor
+            donor["number_don"] += 1
+            donor["avg_don"] = donor["total_don"]/donor["number_don"]
+            print("\nDear {},\n\nThank you for your generous donation of ${:,.2f}.\n\nSincerely,\nThe Best Charity Ever".format(donor["name"],ask_donor))
+            break
+    else: # Check if the entered name is not in the list 
+        ask_donor = float(input("Please enter a donation amount: "))
+        initial_donor_list.append({"name": ask_name,"total_don": ask_donor,"number_don": 1,"avg_don": ask_donor})
+        print("\nDear {},\n\nThank you for your generous donation of ${:,.2f}.\n\nSincerely,\nThe Best Charity Ever".format(ask_name,ask_donor))
+
+def write_letter():
+    """Function to generate letters for each donor in the list with their total donation amount and stores the letters in a new directory."""
+    current = datetime.datetime.now()
+    abs_path = pathlib.Path('./').absolute()
+    final_path = os.path.join(abs_path, "letter_storage/")
+    if not os.path.exists(final_path):
+        os.makedirs(final_path)
+    for line in initial_donor_list:
+        with open("{}{}_{:02}_{:02}_{:02}.txt".format(final_path, line["name"], current.month, current.day, current.year), 'w') as thank_you_file:
+            thank_you_file.write("Dear {},\n\nThank you for your continuous support over the years\nand for your total donation amount of ${:,.2f}.\n\nSincerely,\nThe Best Charity Ever".format(line["name"],line["total_don"]))
+    
+
+if __name__ == '__main__':
+    option_select = {'1': thank_you, '2': list_data, '3': write_letter}
+    while True: # loop until ended by user input '3'
+        user_prompt = input("\nPlease choose from the following menu of actions:\n[1] Send a Thank You\n[2] Create a Report\n[3] Send letters to everyone\n[Press any other key to quit.]\n\nInput: ")
+        if(user_prompt != '1' and user_prompt != '2' and user_prompt != '3'):
```

Comment:
you can also do `if user_prompt not in option_select` 

however I would recommend mapping quit feature to a function that exits the script:
```
def quit_program():
    print("Bye")
    sys.exit()
```